Tests/XcodeGenKitTests/SpecLoadingTests.swift

```diff
@@ -93,15 +93,28 @@ func specLoadingTests() {
                 "source1",
                 ["path": "source2"],
                 ["path": "sourceWithFlags", "compilerFlags": ["-Werror"]],
-                ["path": "sourceWithFlagsStr", "compilerFlags": "-Werror -Wextra"]
+                ["path": "sourceWithFlagsStr", "compilerFlags": "-Werror -Wextra"],
+                ["path": "sourceWithExcludes", "excludes": ["Foo.swift"]],
+                ["path": "sourceWithExcludesStr", "excludes": "Foo.swift Bar.swift"],
+                ["path": "sourceWithCompilerFlagsExcludes", "compilerFlags": ["-Werror"], "excludes": ["Foo.swift"]],
+                ["path": "sourceWithCompilerFlagsExcludesStr", "compilerFlags": "-Werror -Wextra", "excludes": "Foo.swift Bar.swift"],
             ]
             var targetDictionary2 = validTarget
             targetDictionary2["sources"] = "source3"
 
             let target1 = try Target(name: "test", jsonDictionary: targetDictionary1)
             let target2 = try Target(name: "test", jsonDictionary: targetDictionary2)
 
-            try expect(target1.sources) == [Source(path: "source1"), Source(path: "source2"), Source(path: "sourceWithFlags", compilerFlags: ["-Werror"]), Source(path: "sourceWithFlagsStr", compilerFlags: ["-Werror", "-Wextra"])]
+            let target1SourcesExpect = [Source(path: "source1"), 
+                                        Source(path: "source2"), 
+                                        Source(path: "sourceWithFlags", compilerFlags: ["-Werror"]), 
+                                        Source(path: "sourceWithFlagsStr", compilerFlags: ["-Werror", "-Wextra"]),
+                                        Source(path: "sourceWithExcludes", excludes: ["Foo.swift"]), 
+                                        Source(path: "sourceWithExcludesStr", excludes: ["Foo.swift", "Bar.swift"]),
+                                        Source(path: "sourceWithCompilerFlagsExcludes", compilerFlags: ["-Werror"], excludes: ["Foo.swift"]),
+                                        Source(path: "sourceWithCompilerFlagsExcludesStr", compilerFlags: ["-Werror", "-Wextra"], excludes: ["Foo.swift", "Bar.swift"])]
+
+            try expect(target1.sources) == target1SourcesExpect
             try expect(target2.sources) == [Source(path: "source3")]
         }
```

Comment:
It would be great to have some tests in ProjectGeneratorTests under "Sources". You would need to make use of the `getFileReference` function at the bottom, maybe with a new `expectMissingFile` (see `expectFile` for example)

Reply:
@yonaskolb Thanks for the feedback.

I'm absolutely in for having global patterns and in my opinion excluding the whole directory should also be an option.

On a different note, do you also think it makes sense to have `includes` as part of `Source`?

Reply:
I'm not sure it would make sense to have `includes` as part of the Source, as that's what the `path` does. To have multiple sources paths you create multiple sources. But that does raise the question whether excludes should be part of the target and not Source, so that if you have multiple sources you only have to define excludes once. But then again such an approach would probably be used for simply excluding specific file types, and that could be done in the project options level...

Reply:
Fair enough. Since I'm migrating all the my current work's project files to use `Xcodegen` instead, felt it can be a good option, but I agree with you, we can get carried away with it. 

So i've already implemented pattern matching for excludes for both files and directories, have bumped into an issue with one of the test cases `Sources->generates file sources`.

Is including files as a source path supposed to work in general? 

Here's the test case.

```
$0.it("generates file sources") {
                let directories = """
                Sources:
                  A:
                    - a.swift
                    - B:
                      - b.swift
                      - c.jpg
                """
                try createDirectories(directories)

                target.sources = [
                    "Sources/A/a.swift",
                    "Sources/A/B/b.swift",
                    "Sources/A/B/c.jpg",
                ]
                spec.targets = [target]

                let project = try getPbxProj(spec)
                try project.expectFile(paths: ["Sources/A", "a.swift"], names: ["A", "a.swift"], buildPhase: .sources)
                try project.expectFile(paths: ["Sources/A/B", "b.swift"], names: ["B", "b.swift"], buildPhase: .sources)
                try project.expectFile(paths: ["Sources/A/B", "c.jpg"], names: ["B", "c.jpg"], buildPhase: .resources)
            }
```

When this gets resolved I'll push so that you can have a look at the changes.

Reply:
Yeah adding file sources was added recently via #106